showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java

```diff
@@ -0,0 +1,214 @@
+      HttpJsonClientCall<RequestT, ResponseT> call = next.newCall(method, callOptions);
+      return new ForwardingHttpJsonClientCall.SimpleForwardingHttpJsonClientCall<
+          RequestT, ResponseT>(call) {
+        @Override
+        public void start(Listener<ResponseT> responseListener, HttpJsonMetadata requestHeaders) {
+          Listener<ResponseT> forwardingResponseListener =
+              new ForwardingHttpJsonClientCallListener.SimpleForwardingHttpJsonClientCallListener<
+                  ResponseT>(responseListener) {
+                @Override
+                public void onHeaders(HttpJsonMetadata responseHeaders) {
+                  metadata = responseHeaders;
+                  super.onHeaders(responseHeaders);
+                }
+
+                @Override
+                public void onMessage(ResponseT message) {
+                  super.onMessage(message);
+                }
+
+                @Override
+                public void onClose(int statusCode, HttpJsonMetadata trailers) {
+                  super.onClose(statusCode, trailers);
+                }
+              };
+
+          super.start(forwardingResponseListener, requestHeaders);
+        }
+      };
+    }
+  }
+
+  private HttpJsonCapturingClientInterceptor httpJsonInterceptor;
+  private GrpcCapturingClientInterceptor grpcInterceptor;
+  private HttpJsonCapturingClientInterceptor httpJsonComplianceInterceptor;
+  private GrpcCapturingClientInterceptor grpcComplianceInterceptor;
+  private EchoClient grpcClient;
+  private EchoClient httpJsonClient;
+  private ComplianceClient grpcComplianceClient;
+  private ComplianceClient httpJsonComplianceClient;
+
+  @Before
+  public void createClients() throws Exception {
+    // Create gRPC Interceptor and Client
+    grpcInterceptor = new GrpcCapturingClientInterceptor();
+    grpcClient = TestClientInitializer.createGrpcEchoClient(ImmutableList.of(grpcInterceptor));
+
+    // Create HttpJson Interceptor and Client
+    httpJsonInterceptor = new HttpJsonCapturingClientInterceptor();
+    httpJsonClient =
+        TestClientInitializer.createHttpJsonEchoClient(ImmutableList.of(httpJsonInterceptor));
+
+    // Create gRPC ComplianceClient and Interceptor
+    // Creating a compliance client to test case where api version is not set
+    grpcComplianceInterceptor = new GrpcCapturingClientInterceptor();
+    grpcComplianceClient =
+        TestClientInitializer.createGrpcComplianceClient(
+            ImmutableList.of(grpcComplianceInterceptor));
+
+    // Create HttpJson ComplianceClient and Interceptor
+    httpJsonComplianceInterceptor = new HttpJsonCapturingClientInterceptor();
+    httpJsonComplianceClient =
+        TestClientInitializer.createHttpJsonComplianceClient(
+            ImmutableList.of(httpJsonComplianceInterceptor));
+  }
+
+  @After
+  public void destroyClient() {
+    grpcClient.close();
+    httpJsonClient.close();
+    grpcComplianceClient.close();
+    httpJsonComplianceClient.close();
```

Comment:
Can you also add `awaitTermination()` here after the `.close()` calls?

Reply:
IIUC, `.close()` is same as `.shutdown()` ([code](https://github.com/googleapis/sdk-platform-java/blob/8381dc58c3bc0d5821c9c22bb7af49b3ad9e544e/gax-java/gax/src/main/java/com/google/api/gax/core/BackgroundResourceAggregation.java#L93)). We should `.shutdown()` a client or `.awaitTermination()`, not both, and the main difference is that `.awaitTermination()` has timeout ([comment here](https://github.com/googleapis/sdk-platform-java/blob/8381dc58c3bc0d5821c9c22bb7af49b3ad9e544e/gax-java/gax/src/main/java/com/google/api/gax/core/BackgroundResource.java#L70-L74)).
However, if we want to prevent a test hanging, this may not work, as these are called within `@After`  and I believe if a test hangs and does not finish, the `@After` method will not be executed. An alternative is to add `@Test(timeout = ..)` to tests. But I wonder if you have a specific concern about these tests? Or do you prefer to implement timeouts for all integration tests as precautious?

Reply:
We typically use `close()` in conjunction with `awaitTermination()`, with `close()` being called first. You're right `close()` just calls `shutdown()` and all `shutdown()` does to prevent *new* tasks from being scheduled, but still allows currently in-flight tasks to continue. We simply use `awaitTermination()` to allow the current in-flight tasks to finish (it blocks until they are done or until the timeout has been reached).

We had previously seen some grpc-java warnings about managed channels still being allocated without trying to wait to the tasks to terminated. i.e something like:
```
io.grpc.internal.ManagedChannelImpl$ManagedChannelReference cleanQueue
SEVERE: *~*~*~ Channel io.grpc.internal.ManagedChannelImpl-2096 for target 127.0.0.1:8771 was not shutdown properly!!! ~*~*~*
    Make sure to call shutdown()/shutdownNow() and awaitTermination().
java.lang.RuntimeException: ManagedChannel allocation site
```

Reply:
Bumping this.

Reply:
Thanks for explaining. Added.